Panic while selecting a goroutine on Mac OS X

Please answer the following before submitting your issue:

Note: Please include any substantial examples (debug session output,
stacktraces, etc) as linked gists.

1. What version of Delve are you using (`dlv version`)?
```
Delve Debugger
Version: 1.0.0-rc.1
Build:
```
2. What version of Go are you using? (`go version`)?
`go version go1.8.1 darwin/amd64`
3. What operating system and processor architecture are you using?
Mac OS X Sierra
4. What did you do?
Tried to select the goroutine using the command `goroutine 5` (5 = goroutine id)
5. What did you expect to see?
That the goroutine was selected
6. What did you see instead?
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x43042d1]

goroutine 1 [running]:
github.com/derekparker/delve/pkg/terminal.(*Commands).goroutine(0xc42196a8d0, 0xc421972480, 0x0, 0xffffffffffffffff, 0x0, 0x0, 0xc4214520fa, 0x1, 0x142, 0xc421966460)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/pkg/terminal/command.go:508 +0x281
github.com/derekparker/delve/pkg/terminal.(*Commands).(github.com/derekparker/delve/pkg/terminal.goroutine)-fm(0xc421972480, 0x0, 0xffffffffffffffff, 0x0, 0x0, 0xc4214520fa, 0x1, 0x2, 0xb)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/pkg/terminal/command.go:129 +0x7d
github.com/derekparker/delve/pkg/terminal.(*Commands).CallWithContext(0xc42196a8d0, 0xc4214520f0, 0xb, 0xc421972480, 0x0, 0xffffffffffffffff, 0x0, 0x0, 0xb, 0xc4214520f0)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/pkg/terminal/command.go:305 +0x134
github.com/derekparker/delve/pkg/terminal.(*Commands).Call(0xc42196a8d0, 0xc4214520f0, 0xb, 0xc421972480, 0x0, 0x0)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/pkg/terminal/command.go:310 +0x71
github.com/derekparker/delve/pkg/terminal.(*Term).Run(0xc421972480, 0x0, 0x0, 0x0)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/pkg/terminal/terminal.go:131 +0x38f
github.com/derekparker/delve/cmd/dlv/cmds.execute(0x0, 0xc4201481a0, 0x1, 0x1, 0xc4200fd340, 0x0, 0x0, 0x0, 0x0)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/cmd/dlv/cmds/commands.go:507 +0x818
github.com/derekparker/delve/cmd/dlv/cmds.New.func4(0xc420146900, 0xc4201481a0, 0x1, 0x1)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/cmd/dlv/cmds/commands.go:171 +0x6f
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).execute(0xc420146900, 0xc420148150, 0x1, 0x1, 0xc420146900, 0xc420148150)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:647 +0x234
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420146000, 0xc4201466c0, 0xc420146480, 0xc420146240)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:733 +0x339
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420146000, 0xc4200001a0, 0xc4200001a0)
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:692 +0x2b
main.main()
	/Users/renannp/development/go/libraries/src/github.com/derekparker/delve/cmd/dlv/main.go:13 +0x55
```

I was using the default backend (lldb) in mac. After switching to native, that command worked.

--- a/_fixtures/notify-v2.go
+++ b/_fixtures/notify-v2.go
@@ -1,0 +1,23 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+)
+
+func main() {
+	http.HandleFunc("/test", func(w http.ResponseWriter, req *http.Request) {
+		go func() {
+			// I know this is wrong, it is just to simulate a deadlocked goroutine
+			fmt.Println("locking...")
+			mtx := &sync.Mutex{}
+			mtx.Lock()
+			mtx.Lock()
+			fmt.Println("will never print this")
+		}()
+	})
+
+	log.Fatalln(http.ListenAndServe("127.0.0.1:8888", nil))
+}

--- a/pkg/terminal/command.go
+++ b/pkg/terminal/command.go
@@ -473,6 +473,13 @@ func goroutines(t *Term, ctx callContext, argstr string) error {
 	return nil
 }
 
+func selectedGID(state *api.DebuggerState) int {
+	if state.SelectedGoroutine == nil {
+		return 0
+	}
+	return state.SelectedGoroutine.ID
+}
+
 func (c *Commands) goroutine(t *Term, ctx callContext, argstr string) error {
 	args := strings.SplitN(argstr, " ", 2)
 
@@ -505,7 +512,7 @@ func (c *Commands) goroutine(t *Term, ctx callContext, argstr string) error {
 			return err
 		}
 
-		fmt.Printf("Switched from %d to %d (thread %d)\n", oldState.SelectedGoroutine.ID, gid, newState.CurrentThread.ID)
+		fmt.Printf("Switched from %d to %d (thread %d)\n", selectedGID(oldState), gid, newState.CurrentThread.ID)
 		return nil
 	}
 

--- a/pkg/terminal/command_test.go
+++ b/pkg/terminal/command_test.go
@@ -5,12 +5,14 @@ import (
 	"fmt"
 	"io/ioutil"
 	"net"
+	"net/http"
 	"os"
 	"path/filepath"
 	"regexp"
 	"strconv"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/derekparker/delve/pkg/proc/test"
 	"github.com/derekparker/delve/service"
@@ -38,6 +40,8 @@ type FakeTerminal struct {
 	t testing.TB
 }
 
+const logCommandOutput = false
+
 func (ft *FakeTerminal) Exec(cmdstr string) (outstr string, err error) {
 	outfh, err := ioutil.TempFile("", "cmdtestout")
 	if err != nil {
@@ -54,6 +58,9 @@ func (ft *FakeTerminal) Exec(cmdstr string) (outstr string, err error) {
 			ft.t.Fatalf("could not read temporary output file: %v", err)
 		}
 		outstr = string(outbs)
+		if logCommandOutput {
+			ft.t.Logf("command %q -> %q", cmdstr, outstr)
+		}
 		os.Remove(outfh.Name())
 	}()
 	err = ft.cmds.Call(cmdstr, ft.Term)
@@ -579,5 +586,20 @@ func TestCheckpoints(t *testing.T) {
 		listIsAt(t, term, "next", 17, -1, -1)
 		listIsAt(t, term, "next", 18, -1, -1)
 		listIsAt(t, term, "restart c1", 16, -1, -1)
+	})
+}
+
+func TestIssue827(t *testing.T) {
+	// switching goroutines when the current thread isn't running any goroutine
+	// causes nil pointer dereference.
+	withTestTerminal("notify-v2", t, func(term *FakeTerminal) {
+		go func() {
+			time.Sleep(1 * time.Second)
+			http.Get("http://127.0.0.1:8888/test")
+			time.Sleep(1 * time.Second)
+			term.client.Halt()
+		}()
+		term.MustExec("continue")
+		term.MustExec("goroutine 1")
 	})
 }

